### PR TITLE
Add "all field" execution mode to query_string query

### PR DESCRIPTION
**Pull request by dakrone**
This commit introduces a new execution mode for the `query_string` query, which
is intended down the road to be a replacement for the current `_all` field.

It now does auto-field-expansion and auto-leniency when the following criteria
are ALL met:
- The `_all` field is disabled
- No `default_field` has been set in the index settings
- No `default_field` has been set in the request
- No `fields` are specified in the request

Additionally, a user can force the "all-like" execution by setting the
`all_fields` parameter to `true`.

When executing in all field mode, the `query_string` query will look at all the
fields in the mapping that are not metafields and can be searched, and
automatically expand the list of fields that are going to be queried.

Relates to #19784

### CODE REVIEW

**Comment by dakrone** [comment]
I am opening this PR so we can discuss the direction we'd like the "all" query
to go. That's why this PR currently doesn't have super-exhaustive tests or
documentation (though it does have some tests).

For more back story, see: https://github.com/elastic/elasticsearch/issues/19784#issuecomment-251735730

I'd like to get feedback about whether this is the way we want to build this
query, after which I can iterate, make changes, and then polish the
documentation. So any feedback is welcome!

**Comment by dakrone** [comment]
Other features off the top of my head that would be good to **eventually** incorporate:

- changing operator of the top-level bool query (currently hard-coded to `OR` but it would be trivial to add)
- `minimum_should_match`
- perhaps scoring?

**Comment by dadoonet** [comment]
I did not read the code but only the description of the PR and it looks great. I like it!
Just wondering about the naming of this query while we are at it. Does `all` means what we meant?
To me `all` means `match_all` so I think we should try to find another name although it's a way to replace `_all` field.

Something like `everywhere` or similar...

**Comment by dakrone** [comment]
> Just wondering about the naming of this query while we are at it.

The naming is totally up for grabs! I named it "all" because I was planning on punting to get feedback, so other suggestions are totally welcome!

**Comment by markharwood** [comment]
Thanks for the work on this @dakrone 

Working from your description alone I have a couple of points:

I think the use of "cross_fields" is irrelevant here if you're just going to "constant_score" everything. at the end. In this scenario the only thing cross_fields is doing for you is adding a pointless limitation that all fields must share a common analyzer (I think it silently reverts to a "normal" query if this is the case rather than erroring)

I can see we might want to ignore IDF in scoring but I still think `coord` should be a ranking factor (how many of the user's search terms matched). Presumably this just means pushing the constant_score wrapper down to the word-level clauses not the root

Would be nice for us to know more about which of the query string operators (phrase, field, AND/OR) tend to get used by folks in practice.

**Comment by jimferenczi** [comment]
I wonder if we can avoid the new query path and provides this functionality in the existing parser. 
For instance it should be possible to do:

````
{
  "query": {
    "query_string": {
      "query": "foo bar title:baz body:eggplant"
    }
  }
}
````

... and apply your logic on the "foo bar" part only. I am saying this because it would be difficult to handle query_string (and simple_query_string) without a default field which seems to provide a nice OOTB.

 > production 200 10.29.2.1
Expecting to find documents for their production environment where a 200
response code is returned from the given IP address. Since the tokens
are treated individually, all numeric fields (int, long) can be searched
for 200, all the IP address fields can be searched for 10.29.2.1, and
all text-like (string, text, keyword) fields can be searched for
"production". This can closely mimic search the _all field using the
given terms.

I am not sure about the numerics for numerics fields and ip for ip fields restriction. I would prefer another approach, something like create a query on this field if the input is compatible and ignore the part that are not parseable by this field. In such case 200 would also be added to the text-like fields (and the ip as well). 


> Internally, what happens is that the "all" query analyzes the text using
a Lucene analyzer (by default, the whitespace analyzer) to split it
into tokens, so in this case ["foo", "Bar", "baz"] are the tokens
generated.

Since you can group each field depending on the analyzer I think that you could use the analyzer directly for the text fields instead of relying on the whitespace analyzer upfront. For numerics and others you could split on whitespace but it's really counter intuitive to have this split before the analysis takes place (see #20841)


> I think the use of "cross_fields" is irrelevant here if you're just going to "constant_score" everything. at the end. In this scenario the only thing cross_fields is doing for you is adding a pointless limitation that all fields must share a common analyzer (I think it silently reverts to a "normal" query if this is the case rather than erroring)

The `cross_fields` can be useful but if you split on whitespace before the analysis you already have some kind of `cross_fields` behavior (the same than what `query_string`does). The query `foo bar` on two fields would produce:
`(field1:foo OR field2:foo) AND (field1:bar OR field2:bar)`
... with AND as the default operator.

**Comment by dakrone** [comment]
Thanks for the feedback @markharwood & @jimferenczi!

> Would be nice for us to know more about which of the query string operators (phrase, field, AND/OR) tend to get used by folks in practice.

That's a great idea, I am hoping some Kibana users (and others!) chime in here with any information about use cases!

> I wonder if we can avoid the new query path and provides this functionality in the existing parser. For instance it should be possible to ... I am saying this because it would be difficult to handle query_string (and simple_query_string) without a default field which seems to provide a nice OOTB.

I think that is the eventual goal of this, however, I think right now it's more complex, and it'd be good to nail down what we want the "all-ish" query to be before changing the way `query_string` behaves (which may only happen in a major version, while this is additive and could go into 5.1 eventually)

> I am not sure about the numerics for numerics fields and ip for ip fields restriction. I would prefer another approach, something like create a query on this field if the input is compatible and ignore the part that are not parseable by this field. In such case 200 would also be added to the text-like fields (and the ip as well). 

This is what it already does, I apologize, I simplified in my explanation, but it will still search text/keyword fields for "200" and "127.0.0.1", not *only* numeric and IP fields. It only ignores fields that it cannot parse the input text as.

> Since you can group each field depending on the analyzer I think that you could use the analyzer directly for the text fields instead of relying on the whitespace analyzer upfront. For numerics and others you could split on whitespace but it's really counter intuitive to have this split before the analysis takes place (see #20841)

I experimented with doing this at first, but ran into the problem where someone sends the query `foo 200`, which without analysis would try to parse the entire text `"foo 200"` as a numeric to search on a numeric field. This would then not be parseable (it can't parse "foo 200" as a number). As you said though, doing this **only** for numerics might be a better idea that works around this!

> Regarding `cross_fields` ...

I think re-evaluating this (using `cross_fields`) makes sense. I'd like to get more feedback on whether it'd be useful to have scoring (when we talked about it last we talked about doing everything with `constant_score`). Also, determining how this will change if the analysis suggestion Jim made gets applied.

**Comment by jimferenczi** [comment]
I really like how this PR reused the MultiMatchQuery to create the inner queries. 
I don't know if the `analyzer` option is required though. It seems misleading since it is used to parse the query before the analysis so it's more a split analyzer that is not used to create the final field query. Do you have use cases in mind where this analyzer option could be useful ? I think it could be simplified to a simple boolean `split_on_whitespace` or we could simply removed this option. I think it should be removed if the default operator of this query (OR/SHOULD) cannot be changed. 

Regarding scoring I don't think this query should produce a constant score query. Since it uses the `cross_fields` types of the MultiMatchQuery the scoring should be consistent thanks to the blending of the term queries. It would give a nice OOTB experience as opposed to the first match ranked first scenario. Of course some queries will fail regarding the scoring expectation but it should work for simple ones seamlessly.

The code looks good to me, it's simple and well contained. Tough in order to be able to push a first version of this I think we should discuss the following points:
* decide if this query should produce a score or not
* maybe a better name like david proposed
* decide which options should be exposed (default operator, min should match, ...). This point is not required for the first version since we can add the options later.

**Comment by clintongormley** [comment]
To reiterate the conclusions in Prague:

If a user wants to do real full text search against a single field which contains the contents of multiple other fields, they should use `copy_to`.

This leaves the "discovery" use case: _find these search elements regardless of which fields they're in, or what type the field is_.  For this use case, we spoke about adding a new `_all` query which could query all fields in the mapping. We said it was OK to be slower than a query against the `_all` field, and that scoring wasn't important (and should probably be disabled to discourage use of this query for full text search).

The main use case for this `_all` query is to support the query string query.  Today, the query string query tokenizes the query string on whitespace and then passes each token to the underlying field separately (where it might undergo further analysis).  However, this approach is flawed (it breaks n-grams, shingles, and multi-word synonyms) and has been fixed in Lucene 6.2 (see https://issues.apache.org/jira/browse/LUCENE-2605 for details).  There is a PR to expose this in Elasticsearch (https://github.com/elastic/elasticsearch/pull/20965).

That said, the change in LUCENE-2605 is not without problems.  It works well when there is a single default field, but breaks when there are multiple `fields`.  Imagine a query string like `(quick brown fox)~2`.  In the current tokenize-on-whitespace  approach, two of the three tokens must be present in any of the listed `fields`.  With the new approach, two of the three tokens would need to be present in the SAME field (unless we complicate things further by applying the cross-field matching that exists in the `multi_match` query).

The gist of all of the above is that these same issues are shared by the query string query and by the proposed `_all` query.  I'm beginning to think that we shouldn't add the `_all` query, but instead **change the behaviour of the query string query and, when the `default_field` is set to `_all`, expand `fields` to all fields.**

There are three types of queries to deal with:

* **Analyzed `text` field** - you want to pass as much contiguous text as possible to the underlying field to be dealt with by the field's analyzer.  Min-should-match logic could be applied as we do with `cross_fields` matching in `multi_match`.
* **Not analyzed `keyword` field** - Contiguous text should probably be split on white space so that each token can be considered separately.  Keywords can of course contain spaces.  This could be worked around by either (a) splitting on whitespace and using a phrase query for `"keyword with space"` or (b) not splitting on whitespace and relying on  operators  `foo OR bar OR keyword with space`
* **Numeric fields** - Should be split on whitespace and only legal numbers, dates, IP addresses, etc should be passed to the underlying field  (won't work for date formats like `yyyy-mm-dd hh:mm:ss` but could be worked around with phrase queries).

Honestly, the more I write here, the less I like the idea of removing the `_all` field.  There is a helluva lot of complexity involved in trying to do the right thing at query time, and there are lots of cases where we will fail. Explaining the failure is even more complicated.

I'm leaning towards going back to the simpler change of excluding numerics from the `_all` field and being done with it.  It is easy to disable the `_all` field if you don't need/want it.

**Comment by dakrone** [comment]
> change the behaviour of the query string query and, when the `default_field` is set to `_all`, expand `fields` to all fields.

@clintongormley something that concerns me about this is that the `query_string` query can have very expensive queries in it - fuzzy, phrase with large slops, regexp, wildcards (including leading wildcards), etc. If instead of performing these expensive queries on a single field we expanded the query to perform them on dozens (or potentially hundreds) of fields, I worry that it may be something that takes a cluster down.

>  Today, the query string query tokenizes the query string on whitespace and then passes each token to the underlying field separately (where it might undergo further analysis). However, this approach is flawed (it breaks n-grams, shingles, and multi-word synonyms) and has been fixed in Lucene 6.2

I believe this can be handled by Jim's suggestion to have two versions of the query text, an *non-analyzed* version that's used for all the string-ish fields, and a version *split on whitespace* to be used for all the numeric-ish fields.

> Imagine a query string like `(quick brown fox)~2`. In the current tokenize-on-whitespace approach, two of the three tokens must be present in any of the listed fields. With the new approach, two of the three tokens would need to be present in the SAME field

Do we think that's the intended use case for this query? To me this is more of a "match-like" query in that it is for matching, not doing more complicated things. In that case, a dedicated field and potentially using `copy_to` is a better use case. I don't have any hard evidence to back this up though, until more people chime in :)

**Comment by dakrone** [comment]
Regarding the three different types you mentioned @clintongormley, those make sense (I was only thinking about two, but splitting on whitespace for `keyword` fields makes sense alse). If this was done, would you still not want to have this as a separate query?

**Comment by clintongormley** [comment]
> @clintongormley something that concerns me about this is that the query_string query can have very expensive queries in it - fuzzy, phrase with large slops, regexp, wildcards (including leading wildcards), etc. If instead of performing these expensive queries on a single field we expanded the query to perform them on dozens (or potentially hundreds) of fields, I worry that it may be something that takes a cluster down.

But that's essentially what we're talking about doing at the moment. 

> If this was done, would you still not want to have this as a separate query?

I'm not getting how we would support the query string syntax with the `_all` query, without making the `_all` query do everything that the query string query does today (incl fuzzy, min should match etc).

The _all query feels like the wrong answer here.  My fear is it is going to become a big complex mess, while the existing alternative (the `_all` field) is pretty simple.

**Comment by rjernst** [comment]
> I'm leaning towards going back to the simpler change of excluding numerics from the _all field and being done with it. It is easy to disable the _all field if you don't need/want it.

This does not solve these issues (and which is why we have discussed this so much!):
* Kibana would still not be able to do search on "foo 200"
* *everyone* still pays the cost of _all out of the box (essentially doubling the size of their index), when everyone does not actually need this.
* Mappings stay complex (we must keep the `include_in_all` setting, which is invasive internal mapping code)

And there are probably some more things I've already forgotten. We started this discussion because of all the problems with `_all`. Please, let's not just drop this because it is a "big" change. `_all` is broken in many ways right now. Let's fix it.

Regarding some points made above:
* Why can we not keep this simple and leave scoring out of it? I thought that was the consensus before.
* We could handle the numerics by wrapping the analyzer for numeric fields, when used with the all query, with one that splits on whitespace (and discards non-numeric terms before even making it to the numeric parser).
* For (simple_)query_string, can we just document how to create a catch all field with `copy_to`, and require the `fields` parameter (and keep a separate all query or whatever we want to call it, which does not have any special operators)?

**Comment by clintongormley** [comment]
@rjernst what gets exposed to the user?  A simple field whose value gets passed to just the `_all` query?  If so, then it won't be enough - users use the syntax like AND, OR, NOT, ranges, field:, prefix, wildcard etc.  But query string query also supports grouping, fuzzy, min should match, etc.  How do we handle those with the `_all` query?  Build in support for all these features? 

I could imagine changing the query string to auto-expand `_all` to all fields, and adding some of the things I've mentioned, but the `_all` query doesn't feel like the answer here.

**Comment by dakrone** [comment]
> A simple field whose value gets passed to just the _all query? If so, then it won't be enough - users use the syntax like AND, OR, NOT, ranges, field:, prefix, wildcard etc. But query string query also supports grouping, fuzzy, min should match, etc. How do we handle those with the _all query?

I think we're conflating two different use cases. I would argue that for the users that are advanced enough to use prefixes, wildcards, regexp, fuzzy, etc that the "all" query is not for them. These users can use the `:` syntax to specify which fields they want to match on in their query.

To me, the "all" query is for the users that have no idea what that search box in Kibana is backed by, and simply want easy-to-use filtering functionality that works without knowing anything about their data, the mappings, and without fear of `query_string`'s error messages.

I have pushed a change that separates the query text into an analyzed and non-analyzed version (as people have recommended) and uses the appropriate version depending on the field type. This allows a `keyword` field with whitespace to match.

**Comment by dakrone** [comment]
> maybe a better name like david proposed

If we get to this point, maybe `"all_fields"` is a better query name?

**Comment by clintongormley** [comment]
> I think we're conflating two different use cases. I would argue that for the users that are advanced enough to use prefixes, wildcards, regexp, fuzzy, etc that the "all" query is not for them. These users can use the : syntax to specify which fields they want to match on in their query.

Think about Kibana users.  Sure, some use the search bar just for filtering.  Others use it with at least ranges, wildcards, AND|OR|NOT etc.  Others use a lot more.  Either way, there's only one search field exposed in Kibana. So either we're exposing the query string query, or we're coming up with a new query parser syntax that can be backed by this _all query.

**Comment by jimferenczi** [comment]
> Either way, there's only one search field exposed in Kibana.

I am confused here, Kibana exposes the `query_string` query as is and the default field is the one defined in the targeted ES, right ? 

> So either we're exposing the query string query, or we're coming up with a new query parser syntax that can be backed by this _all query.

I agree with Clinton, why can't we just change the logic when _all is the default field for a `query_string` query ? This parser is easily modifiable and we can tweak everything we want in it. 

In fact this simple query:
````
GET t/t/_validate/query?explain
{
    "query": {
        "query_string": {
           "query": "this is a test",
           "fields": ["*"],
           "lenient": true
        }
    }
}
````
... is really close to what the proposed _all query is doing. The only difference I see is that the `query_string` does not filter the metadata fields. This would be much simpler than adding the required syntax to a new _all query parser.

**Comment by dadoonet** [comment]
I agree with @jimferenczi and @clintongormley:

> I wonder if we can avoid the new query path and provides this functionality in the existing parser. 
> For instance it should be possible to do:
> 
```
{
  "query": {
    "query_string": {
      "query": "foo bar title:baz body:eggplant"
    }
  }
}
```
> ... and apply your logic on the "foo bar" part only. I am saying this because it would be difficult to handle query_string (and simple_query_string) without a default field which seems to provide a nice OOTB.

So no need for another query type if we can get a nice OOTB without the need of generating a `_all` field at index time.

**Comment by dakrone** [comment]
> In fact this simple query `...` is really close to what the proposed _all query is doing. The only difference I see is that the query_string does not filter the metadata fields. This would be much simpler than adding the required syntax to a new _all query parser.

I wish it did, however, if what @clintongormley said about users using `AND` and `OR` is correct, the `AND` functionality is broken in this query: https://gist.github.com/dakrone/022cf20c8e390c4063aa263cd9370e5a which is an unfortunate thing about the `query_string` query.

**Comment by dakrone** [comment]
(By the way I'm not against re-using the `query_string` query for this functionality, I'm still looking into it. I don't want anyone to think I'm against it and trying to stop the conversation)

**Comment by dakrone** [comment]
Found it. The unintended behavior is related to bools having an implicit `boolfield:true` clause in the query. I'll look at how hard it would be to change this for `query_string`.

**Comment by dakrone** [comment]
Okay, I have made some changes to the `query_string` query to get this sort of behavior. It now does auto-field-expansion and auto-leniency when the following criteria are *ALL* met:

- The `_all` field is disabled
- No `default_field` has been set in the index settings
- No `default_field` has been set in the request
- No `fields` are specified in the request

It works great so far! What do we think of this as a solution? 

This would allow Kibana to retain almost all of its behavior without any changes other than disabling the `_all` field by default on the ES side.

File: core/src/test/java/org/elasticsearch/search/query/QueryStringIT.java
Comment (jimferenczi) [inline core/src/test/java/org/elasticsearch/search/query/QueryStringIT.java]: queryStringQuery("Foo Bar").splitOnWhitespace(false) ?

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java
Comment (jimferenczi) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: Maybe better to use MapperService.isMetadataField since "_" is a valid prefix for a user field name.

**Review by jimczi — COMMENTED** [review]
This looks great ! 
Left two minors, are you going to remove the all query ? 
I am not sure how we should handle split_on_whitespace=false with this new expansion. 
For instance the query "foo 3" would produce a term query for "3" on a number field but if split_on_whitespace=false this would result in an empty query for the same number field. I think we should try to split on whitespace for all non-tokenized field in order to make the behavior consistent ? Though as I said it could be done in a follow up when https://github.com/elastic/elasticsearch/pull/20965 is in.

File: core/src/test/java/org/elasticsearch/search/query/QueryStringIT.java
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/search/query/QueryStringIT.java]: I've changed the commented out test to add this, just waiting for the PR to be merged now (hint hint! ;) )

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: That's a better way to do this, I've changed this line.

**Comment by dakrone** [comment]
Thanks for the feedback! I've pushed a couple of commits addressing your feedback.

> are you going to remove the all query ?

I definitely can, I left it in just because the `query_string` change was additive, so it's really easy to pull the "all" query out now assuming we're happy with the logic in `query_string`.

Are we happy with it? What's the consensus here?

> I am not sure how we should handle split_on_whitespace=false with this new expansion. I think we should try to split on whitespace for all non-tokenized field in order to make the behavior consistent? Though as I said it could be done in a follow up when #20965 is in.

I'm not sure how to handle it yet either, I'm totally up for punting on this until the feature is in, and then we can revisit how we want to handle it for this case.

**Comment by dakrone** [comment]
I pushed a commit to remove the "all" query, since it's a separate commit on this PR I can always revert it if we want it back.

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by dakrone** [comment]
@jimczi I pushed another commit for this that adds the `all_fields` parameter to force execution using `query_string`'s "auto-all" mode. This will help anyone that wants to compare a query against the `_all` field to a query against all automatic fields. It also adds some documentation about the new behavior of `query_string`. Can you take another look?

**Comment by jimczi** [comment]
Great, so currently the all_fields mode is activated if:
- `_all` field is disabled and `index.query.default_field` has not been changed from its default (`_all`)

- `all_fields` is explicitly set in the query. 

I understand that we need a way to force this mode but we should not need another parameter for this ? Maybe we could instead create a new metadata field named `_all_fields` that would refer to all fields in the mapping:

````
POST /test/doc/_search?pretty
{
  "query": {
    "query_string": {
      "query": "secarity~1 f_int:200 127.0.0.1 \"2016/09/01\"",
      "default_operator": "AND",
      "default_field": "_all_fields"
    }
  }
}
````

I think it would be easier to replicate this mode in the other full text queries if we name it consistently and reuse the parameters that are already present ?

**Comment by dakrone** [comment]
> I understand that we need a way to force this mode but we should not need another parameter for this ? Maybe we could instead create a new metadata field named `_all_fields` that would refer to all fields in the mapping
> I think it would be easier to replicate this mode in the other full text queries if we name it consistently and reuse the parameters that are already present ? 

I think a separate parameter is better, because then there is no ambiguity between which field supports this execution style and which does not. For example `{"term": {"_all_fields": "foo bar baz"}}`, which is strange because a term query is expected to execute on a single field. So a user that expects `_all_fields` to work everywhere is going to find it works for some queries and doesn't for others.

While a new user can go look at their mappings to see what is configured for the `_all` field, they would never be able to find `_all_fields` in their mappings, which I think would lead to more confusion.

Additionally, if we create a meta field for this, it complicates boosting parsing, so for example:

```
GET /_search
{
  "query": {
    "multi_match" : {
      "query":    "this is a test", 
      "fields": [ "_all_fields^2.3", "message" ] 
    }
  }
}
```

I think this will not do what the user intends it to do. So, I'm in favor of the explicit `all_fields` parameter that we can add to whichever queries we want to support this on in the future.

**Comment by jimczi** [comment]
> So, I'm in favor of the explicit all_fields parameter that we can add to whichever queries we want to support this on in the future.

I am not sure of the magic `_all_fields` fields either so I am ok for this option.
Though can you make it mutually exclusive of the `default_field` parameter ? That's what confused me when I first looked at it.

**Comment by dakrone** [comment]
Sure, I'll change it so it throws an exception if you specify both options.
Will ping you when it's added.

**Comment by dakrone** [comment]
@jimczi okay, I pushed a commit that will prevent `all_fields` with being used with `default_field` or `fields` (and a test)

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: I think that the Legacy field mappers are missing (at least those that are stil valid in 2.x ?).

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: The legacy field mappers share the same `CONTENT_TYPE` constant values as all the values in `NumberFieldMapper.NumberType.values()`.

ie, we don't need `LegacyLongFieldMapper` because `LegacyLongFieldMapper.CONTENT_TYPE` is `"long"` and we already add `NumberFieldMapper.NumberType.LONG`, which is `"long"`

**Review by jimczi — COMMENTED** [review]
@dakrone I left one comment regarding the accepted field types.

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: Ah CONTENT_TYPE I see. Sorry for the noise ;)

**Review by jimczi — APPROVED** [review]
Phrase queries may be problematic when some of the searchable fields have their positions disabled. 
We should probably handle this case without failing but this can be done in a follow up. 
I am sure there are other corner cases that we'll discover later which is why we should merge this PR and start to play with it ;).
LGTM

**Comment by dakrone** [comment]
@clintongormley do you have any more reservations before this PR is merged?

**Comment by tylerfontaine** [comment]
I see this is tagged for 5.1.0 - are we not considering this a breaking change? Or will the defaults be unchanged for 5.1, just the ability to use this if desired

**Comment by Bargs** [comment]
Sorry if I missed this in the comments, but what happens if a search is executed against multiple indices and the `_all` field is disabled in some, but not others? Same question for `default_field`. Does auto-expansion still happen for all indices being searched? If `_all` is disabled by default for new indices, I imagine this scenario will be pretty common for Kibana users issuing searches against index patterns.

**Comment by dakrone** [comment]
> what happens if a search is executed against multiple indices and the `_all` field is disabled in some, but not others? Same question for `default_field`. Does auto-expansion still happen for all indices being searched?

The auto-execution is per-shard, so for the shards that have `_all` disabled they'll do the auto-expand, for the ones that have `_all`, it'll search on `_all` like it does now.

**Comment by Bargs** [comment]
Great, I was hoping it was something like that. That way the results of a user's existing saved searches won't change out from underneath them.

**Comment by dakrone** [comment]
> That way the results of a user's existing saved searches won't change out from underneath them.

@Bargs it's important to note that with this change, nothing will be different right now, this still leaves `_all` enabled by default, so until we make a change there, this would require a user to change their template in order to see this in Kibana

**Comment by clintongormley** [comment]
LGreatTM!

**Comment by dakrone** [comment]
Thanks everyone, for the reviews!

**Comment by jpountz** [comment]
Should we try to consolidate the `all_fields: true` and `fields: [ "*" ]` options? From a user perspective, they look like they are achieving the same thing?

**Comment by anhlqn** [comment]
I've just upgraded a test cluster from ES and Kibana from 2.4 to 5.6.8 recently, and I am grateful for this change in Log Analysis use case. In ES 2.4, to provide end users with an easy Discovery mode when searching syslog events, I set `include_in_all: false` at index level and set `include_in_all: true` for some fields. That approach allows users to make initial full text searches without me blowing up index size for leaving `_all` field enabled for all fields. 

The biggest field for syslog events in my case is the `message` field, and to save storage, I've set it to
```
"message": {
  "type": "string",
  "index": "no",
  "include_in_all": "true"
}
``` 
With the discontinue of `_all` field since ES v6, it seems that to preserve the same search experience, I have two choices:

1. Change `message` field type to `text` and reindex existing data to prevent mapping conflict (1 year of data at tens of TB)
2. Make use of `copy_to` to copy `message` to another field, perhaps to the `all_fields`
```
"message": {
  "type": "text",
  "index": false,
  "copy_to": "all_fields"
}
```
My question is that is number 2 a good approach and is there any other ways?

**Comment by dakrone** [comment]
Hi @anhlqn, this sounds like a good questions for the discuss forums at https://discuss.elastic.co as we reserve Github issues for feature requests and bug reports. If you want to ask your question there it is a much better place to discuss things like this.

**Pull request merged**
